lm4tools: init at 0.1.3

rrbutani opened the pull request:
###### Description of changes

Adds [`lm4tools`](https://github.com/utzig/lm4tools), a pair of tools for embedded development on the [TI Stellaris Launchpad](https://www.ti.com/tool/EK-LM4F120XL) and derivatives, as a package.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
  - [ ] x86_64-linux `pkgsStatic`
  - [ ] aarch64-linux `pkgsStatic`
  - [ ] `pkgsCross.mingwW64`, static (from an `x86_64-linux` builder)
  - [ ] x86_64-darwin static (with `stdenvAdapters.makeStatic` as described [here](https://github.com/ut-utp/.github/wiki/lm4flash-Binaries))
  - [ ] aarch64-darwin static (with `stdenvAdapters.makeStatic` as described [here](https://github.com/ut-utp/.github/wiki/lm4flash-Binaries))
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`?
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rrbutani committed:
lm4tools: init at 0.1.3

rrbutani committed:
libusb1: fixes for building for windows

`nix build .#pkgsCross.mingwW64.libusb1`

rrbutani committed:
lm4tools: fixes for cross-compilation and windows

rrbutani commented:
While trying to get `lm4tools` to compile for `minGW64`, I ran into a build error for `libusb1` (`lm4tools`' only dep) when building for Windows (i.e. `np.pkgsCross.mingwW64.libusb1`); I've included [a fix](https://github.com/NixOS/nixpkgs/pull/175052/commits/9f3606ea12b2e003771b546b8ec0232e6ace8a5a) for that build error on this branch but I can spin that off into its own PR if required.

rrbutani committed:
lm4tools: fix static compilation on macOS

FliegendeWurst commented:
Please rebase your PR to target staging (see https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging).

prusnak commented:
I wonder why this PR causes so many rebuilds while the change in libusb1 should affect only windows 🤔

FliegendeWurst commented:
> I wonder why this PR causes so many rebuilds while the change in libusb1 should affect only windows :thinking:

It also sets `dontAddDisableDepTrack = false;` on every other platform (previous value: `null`).

rrbutani commented:
My bad, sorry about that!

---

@FliegendeWurst do you still want me to rebase?

FliegendeWurst commented:
No longer necessary. I like your `lib.optionalAttrs` solution :)

doronbehar commented:
@rrbutani do you still want to work on this? It still needs a lot of work.

rrbutani committed:
lm4tools: use `hostPlatform.emulator` for tests

rrbutani committed:
lm4tools: mark as broken on Windows when *not* doing a static build

This:
```nix
(import ./. {
    crossSystem = (import ./. {}).lib.systems.examples.mingwW64
        // { isStatic = true; }
    ;
}).lm4tools.tests
```
is fine

but this is not:
```nix
(import ./. {}).pkgsCross.mingwW64.lm4tools.tests
```

For the latter wine complains with:
```
0024:err:module:import_dll Library libusb-1.0.dll (which is needed by L"Z:\\nix\\store\\q07favlimqpmi48g48bbw0wyy1fvn2ar-lm4tools-x86_64-w64-mingw32-0.1.3\\bin\\lm4flash.exe") not found
0024:err:module:LdrInitializeThunk Importing dlls for L"Z:\\nix\\store\\q07favlimqpmi48g48bbw0wyy1fvn2ar-lm4tools-x86_64-w64-mingw32-0.1.3\\bin\\lm4flash.exe" failed, status c0000135
```

---

win-dll-link is supposed to take care of symlinking `libusb1.dll` for
us; the issue is that the `libusb1` build does not seem to produce a DLL

doronbehar commented on pkgs/development/embedded/lm4tools/default.nix:
```diff
@@ -0,0 +1,114 @@
+
+  patches = [
+    # We don't want the Makefile to assume `libusb` lives in /usr/local/lib.
+    ./macOS-use-pkg-config.patch
+    # When cross compiling `pkg-config` may have a prefix; we want to get its
+    # name/path from `$PKG_CONFIG`.
+    ./use-pkg-config-env-var.patch
+    # Build with `-O2`:
+    ./use-release-config-for-build.patch
+    # Have the `Makefile`s pass `pkg-config` the right flags when doing a
+    # static build.
+    #
+    # See the comment on `configurePhase` below.
+    ./pkg-config-static-build-support.patch
+  ] ++ lib.optionals stdenv.hostPlatform.isWindows [
+    # The minGW toolchain appends a `.exe` if it's not present which confuses
+    # `install`.
+    ./windows-file-ext.patch
+  ] ++ lib.optionals (!buildLmicdiusb) [
+    ./disable-lmicdiusb-build.patch
+  ];
```
Have you tried to upstream all / some of these patches?

doronbehar commented on pkgs/development/embedded/lm4tools/default.nix:
```diff
@@ -0,0 +1,114 @@
+  # So, we look for the `--enable-static` configure flag as an indicator
+  # that we're linking statically (even though the `makeStaticDarwin` adapter
+  # in isolation does *not* add this configure flag):
+  configurePhase = ''
+    if [[ "$configureFlags" =~ "--enable-static" ]]; then
+      makeFlagsArray+=(STATIC_BUILD=1)
+    fi
+  '';
```
This also looks like something that should be upstreamed.

doronbehar commented on pkgs/development/embedded/lm4tools/default.nix:
```diff
@@ -0,0 +1,114 @@
+    };
+  };
+
+  meta = with lib; rec {
+    description  = "Tools for TI Stellaris boards";
+    longDescription = ''
+      Tools to enable multi-platform development on the TI Stellaris
+      Launchpad boards (lm4f) and TI Tiva C boards (tm4c).
+
+      Includes `lm4flash` and `lmicdiusb`.
+    '';
+    homepage     = "https://github.com/utzig/lm4tools";
+    downloadPage = "${homepage}/releases/tag/v${finalAttrs.version}";
+    changelog    = downloadPage;
```
Nit: How about removing `rec`, and using:

```nix
  inherit (src.meta) homepage;
  downloadPage = "${src.meta.homepage}/releases/tag/v${finalAttrs.version}";
```

doronbehar commented on pkgs/development/embedded/lm4tools/default.nix:
```diff
@@ -0,0 +1,114 @@
+    maintainers  = with lib.maintainers; [ rrbutani ];
+    mainProgram  = "lm4flash";
+    platforms    = with lib.platforms; unix ++ windows;
+    badPlatforms = with lib.systems.inspect; [
+      # `libusb1` doesn't appear to produce DLLs; we get runtime errors when
+      # not linking statically.
+      { isStatic = false; parsed = patterns.isWindows; }
```
Interesting! Are you sure this produces a valid Nix list of badPlatforms?

doronbehar commented on pkgs/development/embedded/lm4tools/default.nix:
```diff
@@ -0,0 +1,114 @@
+    };
+  };
+
+  meta = with lib; rec {
```
Considering the extensive use of `with lib...` down here, I'd let go of this `with lib;`

doronbehar commented:
Also, one more thing: please put the `libusb1` commit first, and then squash all other `lm4tools` commits into a single `lm4tools: init at ...` commit.
